routing/index_graph_starter.cpp

```diff
@@ -1,180 +1,396 @@
+    auto const it = m_fake.m_realToFake.find(edge.GetTarget());
+    if (it == m_fake.m_realToFake.end())
+      continue;
+    for (auto const & s : it->second)
+    {
+      // Check fake segment is connected to source segment.
+      if (GetJunction(s, false /* front */ ) == GetJunction(segment, true) ||
+          GetJunction(s, true) == GetJunction(segment, false))
+      edges.emplace_back(s, RouteWeight(edge.GetWeight()));
+    }
   }
+}
 
-  if (m_graph.GetMode() == WorldGraph::Mode::LeapsOnly && (m_start.Fits(segment) || m_finish.Fits(segment)))
+void IndexGraphStarter::AddRealEdges(Segment const & segment, bool isOutgoing,
+                                     std::vector<SegmentEdge> & edges) const
+{
+  if (m_graph.GetMode() == WorldGraph::Mode::LeapsOnly &&
+      m_fake.m_realToFake.find(segment) != m_fake.m_realToFake.end())
   {
     ConnectLeapToTransitions(segment, isOutgoing, edges);
     return;
   }
-
+  
   m_graph.GetEdgeList(segment, isOutgoing, IsLeap(segment.GetMwmId()), edges);
-  GetNormalToFakeEdge(segment, m_start, kStartFakeSegment, isOutgoing, edges);
-  GetNormalToFakeEdge(segment, m_finish, kFinishFakeSegment, isOutgoing, edges);
 }
 
-void IndexGraphStarter::GetFakeToNormalEdges(FakeVertex const & fakeVertex, bool isOutgoing,
-                                             vector<SegmentEdge> & edges)
+void IndexGraphStarter::GetEdgesList(Segment const & segment, bool isOutgoing,
+                                     vector<SegmentEdge> & edges) const
 {
-  if (fakeVertex.GetStrictForward())
-  {
-    GetFakeToNormalEdge(fakeVertex, fakeVertex.GetSegment().IsForward(), edges);
-  }
-  else
+  edges.clear();
+  if (IsFakeSegment(segment))
   {
-    GetFakeToNormalEdge(fakeVertex, true /* forward */, edges);
+    auto const vertexIt = m_fake.m_segmentToVertex.find(segment);
+    CHECK(vertexIt != m_fake.m_segmentToVertex.end(),
+          ("Can not map fake segment", segment, "to fake vertex."));
 
-    if (!m_graph.GetRoadGeometry(fakeVertex.GetMwmId(), fakeVertex.GetFeatureId()).IsOneWay())
-      GetFakeToNormalEdge(fakeVertex, false /* forward */, edges);
-  }
-}
+    if (vertexIt->second.GetType() == FakeVertex::Type::PartOfReal)
+    {
+      auto const realIt = m_fake.m_fakeToReal.find(segment);
+      CHECK(realIt != m_fake.m_fakeToReal.end(),
+            ("Can not map fake segment", segment, "to real segment."));
+      AddRealEdges(realIt->second, isOutgoing, edges);
+    }
 
-void IndexGraphStarter::GetFakeToNormalEdge(FakeVertex const & fakeVertex, bool forward,
-                                            vector<SegmentEdge> & edges)
-{
-  auto const segment = fakeVertex.GetSegmentWithDirection(forward);
-  m2::PointD const & pointTo = GetPoint(segment, true /* front */);
-  RouteWeight const weight(m_graph.GetEstimator().CalcLeapWeight(fakeVertex.GetPoint(), pointTo));
-  edges.emplace_back(segment, weight);
-}
+    auto const & fakeGraph = isOutgoing ? m_fake.m_outgoing : m_fake.m_ingoing;
```

Comment:
It's not a fake graphs, it's more like adjacent edges lists or something like that.

Reply:
done